{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Global scope\n",
    "baseDir = \"/scratch/etl-cartik/dbGaP2I2b2TM/etl-client/data/CONSENT/\"\n",
    "StartDF = pd.DataFrame(columns=['unique_subject_key', 'SUBJECT_ID', 'dbGaP_Subject_ID', 'topmed_study', 'CONSENT', 'SOURCE_SUBJECT_ID'])\n",
    "\n",
    "#1\n",
    "CopdConsentFileName1 = 'phs000179.v6.pht002237.v3.p2.COPDGene_Subject.MULTI.txt'\n",
    "CopdConsentFileName2 = 'phs000951.v3.pht005050.v3.p3.TOPMed_WGS_COPDGene_Subject.MULTI.txt'\n",
    "\n",
    "#2\n",
    "FhsConsentFileName1 = 'phs000007.v30.pht000182.v13.p11.Framingham_Subject.MULTI.txt'\n",
    "FhsConsentFileName2 = 'phs000974.v3.pht004909.v2.p2.TOPMed_WGS_FHS_Subject.MULTI.txt'\n",
    "\n",
    "#3\n",
    "MesaConsentFileName = 'phs000209.v13.pht001108.v4.p3.MESA_Subject.MULTI.txt'\n",
    "\n",
    "#4\n",
    "JhsConsentFileName1 = 'phs000286.v6.pht001920.v5.p2.JHS_Subject.MULTI.txt'\n",
    "JhsConsentFileName2 = 'phs000964.v3.pht004838.v1.p1.TOPMed_WGS_JHS_Subject.MULTI.txt'\n",
    "\n",
    "#5\n",
    "AmishConsentFileName = 'phs000956.v3.pht005000.v1.p1.TOPMed_WGS_Amish_Subject.MULTI.txt'\n",
    "\n",
    "#6\n",
    "ChsConsentFileName = 'phs000287.v6.pht001447.v4.p1.CHS_Subject.MULTI.txt'\n",
    "\n",
    "#7\n",
    "WhiConsentFileName = 'phs000200.v11.pht000982.v8.p3.WHI_Subject.MULTI.txt'\n",
    "\n",
    "#8\n",
    "HvhConsentFileName1 = 'phs000993.v3.pht005013.v2.p2.TOPMed_WGS_HVH_Subject.MULTI.txt'\n",
    "HvhConsentFileName2 = 'phs001013.v3.pht005309.v3.p2.HVH_Subject.MULTI.txt'\n",
    "\n",
    "#9\n",
    "AricConsentFileName1 = 'phs000280.v5.pht001440.v5.p1.ARIC_Subject.MULTI.txt'\n",
    "AricConsentFileName2 = 'phs001211.v2.pht005754.v2.p2.TOPMed_WGS_ARIC_Subject.MULTI.txt'\n",
    "\n",
    "#10\n",
    "CardiaConsentFileName = 'phs000285.v3.pht001554.v3.p2.CARDIA_Cohort_Subject.MULTI.txt'\n",
    "\n",
    "#11\n",
    "CfsConsentFileName = 'phs000954.v2.pht006297.v1.p1.TOPMed_WGS_CFS_Subject.MULTI.txt'\n",
    "\n",
    "#12\n",
    "VteConsentFileName = 'phs000289.v2.pht001883.v1.p1.Venous_Thrombosis_Subject.MULTI.txt'\n",
    "\n",
    "#13\n",
    "CraConsentFileName = 'phs000988.v3.pht005245.v3.p1.TOPMed_WGS_Asthma_Costa_Rica_Subject.MULTI.txt'\n",
    "\n",
    "#14\n",
    "GoldnConsentFileName = 'phs000741.v2.pht003915.v2.p1.GOLDN_Lipid_Lowering_Subject.MULTI.txt'\n",
    "\n",
    "#15 \n",
    "HchsConsentFileName = 'phs000810.v1.pht004713.v1.p1.HCHS_SOL_Cohort_Subject.MULTI.txt'\n",
    "\n",
    "#16 \n",
    "SasConsentFileName1 = 'phs000914.v1.pht005251.v1.p1.Samoans_Adiposity_Subject.MULTI.txt'\n",
    "SasConsentFileName2 = 'phs000972.v4.pht005679.v1.p1.TOPMed_WGS_Samoans_Adiposity_Subject.MULTI.txt'\n",
    "\n",
    "#17 \n",
    "GenoaConsentFileName = 'phs001238.v2.pht006028.v2.p1.GENOA_Subject.MULTI.txt'\n",
    "\n",
    "DbGapParamName = 'dbGaP_Subject_ID'\n",
    "SubjectIdParamName = 'SUBJECT_ID'\n",
    "ConsentParamName = 'CONSENT'\n",
    "SSIdParamName = 'SOURCE_SUBJECT_ID'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##READ IN ALL THE DATA FROM THE HARMONIZED FILES FROM DataSTAGE DCC AND GENERATE ONE MONOLITHIC FILE\n",
    "def GenerateHarmonizedDF():\n",
    "    hsubdir1 = \"DCC_Harmonized/topmed_dcc_harmonized_atherosclerosis_v1/\"\n",
    "    hsubdir2 = \"DCC_Harmonized/topmed_dcc_harmonized_baseline_common_covariates_v2/\"\n",
    "    hsubdir3 = \"DCC_Harmonized/topmed_dcc_harmonized_blood_cell_count_v3/\"\n",
    "    hsubdir4 = \"DCC_Harmonized/topmed_dcc_harmonized_blood_pressure_v1/\"\n",
    "    hsubdir5 = \"DCC_Harmonized/topmed_dcc_harmonized_demographic_v3/\"\n",
    "    hsubdir6 = \"DCC_Harmonized/topmed_dcc_harmonized_lipids_v3/\"\n",
    "    hsubdir7 = \"DCC_Harmonized/topmed_dcc_harmonized_vte_v1/\"\n",
    "    \n",
    "    fname1 = \"topmed_dcc_harmonized_atherosclerosis_v1_eav.txt\"\n",
    "    fname2 = \"topmed_dcc_harmonized_baseline_common_covariates_v2_eav.txt\"\n",
    "    fname3 = \"topmed_dcc_harmonized_blood_cell_count_v3_eav.txt\"\n",
    "    fname4 = \"topmed_dcc_harmonized_blood_pressure_v1_eav.txt\"\n",
    "    fname5 = \"topmed_dcc_harmonized_demographic_v3_eav.txt\"\n",
    "    fname6 = \"topmed_dcc_harmonized_lipids_v3_eav.txt\"\n",
    "    fname7 = \"topmed_dcc_harmonized_vte_v1_eav.txt\"\n",
    "    \n",
    "    hfile1 = baseDir + hsubdir1 + fname1\n",
    "    hfile2 = baseDir + hsubdir2 + fname2\n",
    "    hfile3 = baseDir + hsubdir3 + fname3\n",
    "    hfile4 = baseDir + hsubdir4 + fname4\n",
    "    hfile5 = baseDir + hsubdir5 + fname5\n",
    "    hfile6 = baseDir + hsubdir6 + fname6\n",
    "    hfile7 = baseDir + hsubdir7 + fname7\n",
    "    \n",
    "    AtherosclerosisDF = pd.read_csv(hfile1, comment=\"#\", sep=\"\\t\")\n",
    "    CommonCovariatesDF = pd.read_csv(hfile2, comment=\"#\", sep=\"\\t\")\n",
    "    BloodCellCountDF = pd.read_csv(hfile3, comment=\"#\", sep=\"\\t\")\n",
    "    BloodPressureDF = pd.read_csv(hfile4, comment=\"#\", sep=\"\\t\")\n",
    "    DemographicDF = pd.read_csv(hfile5, comment=\"#\", sep=\"\\t\")\n",
    "    LipidsDF = pd.read_csv(hfile6, comment=\"#\", sep=\"\\t\")\n",
    "    VteDF = pd.read_csv(hfile7, comment=\"#\", sep=\"\\t\")\n",
    "    \n",
    "    HarmonizedDF = pd.concat([AtherosclerosisDF, CommonCovariatesDF])\n",
    "    HarmonizedDF = pd.concat([HarmonizedDF, BloodCellCountDF])\n",
    "    HarmonizedDF = pd.concat([HarmonizedDF, BloodPressureDF])\n",
    "    HarmonizedDF = pd.concat([HarmonizedDF, DemographicDF])\n",
    "    HarmonizedDF = pd.concat([HarmonizedDF, LipidsDF])\n",
    "    HarmonizedDF = pd.concat([HarmonizedDF, VteDF])\n",
    "    return HarmonizedDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "##HELPER METHOD TO ASSEMBLE A DATA FRAME OF A STUDY SUBSET FROM THE HARMONIZED DATASETS\n",
    "def GetStudySubsetFromHarmonizedData(HarmonizedDF, StudyName, Int64Flag=False):\n",
    "    SubsetDF = HarmonizedDF.loc[HarmonizedDF['topmed_study'] == StudyName]\n",
    "    SubsetDF = SubsetDF[[SubjectIdParamName, 'unique_subject_key', 'topmed_study']]\n",
    "    SubsetDF = SubsetDF.drop_duplicates()\n",
    "    print str(len(pd.unique(SubsetDF[SubjectIdParamName]))) + ' patients\\' data has been harmonized from ' + StudyName + ' study'\n",
    "    if Int64Flag:\n",
    "        SubsetDF = SubsetDF.astype({'SUBJECT_ID': 'int64'}) ##Because SUBJECT_ID in CONSENT Tables is int64\n",
    "    return SubsetDF    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "##HELPER METHOD TO GENERATE CONSENT DATA FRAME. DEFAULT VERSION \n",
    "def ProcessConsentFile(ConsentFileName, StudyName, ColumnToUseForSourceSubjectId=SSIdParamName):\n",
    "    ConsentFile = baseDir + 'data_latest_version/' + ConsentFileName\n",
    "    ConsentDF = pd.read_csv(ConsentFile, comment=\"#\", sep=\"\\t\")\n",
    "    \n",
    "    if(ColumnToUseForSourceSubjectId != SSIdParamName):\n",
    "        ConsentDF[SSIdParamName] = ConsentDF[ColumnToUseForSourceSubjectId]\n",
    "    ConsentDF = ConsentDF[[DbGapParamName, SubjectIdParamName, ConsentParamName, SSIdParamName]]\n",
    "    ConsentDF = ConsentDF.drop_duplicates()\n",
    "    \n",
    "    print str(len(pd.unique(ConsentDF[SubjectIdParamName]))) + ' ' + StudyName + ' patients in consent file'\n",
    "    \n",
    "    return ConsentDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "##HELPER METHOD TO GENERATE CONSENT DATA FRAME WHERE SOME COLUMN RENAMING IS REQUIRED\n",
    "def ProcessConsentFileWithColumnRenaming(ConsentFileName, StudyName, ColumnsToRename, ColumnToUseForSourceSubjectId=SSIdParamName):\n",
    "    ConsentFile = baseDir + 'data_latest_version/' + ConsentFileName\n",
    "    ConsentDF = pd.read_csv(ConsentFile, comment=\"#\", sep=\"\\t\")\n",
    "    \n",
    "    if(ColumnToUseForSourceSubjectId != SSIdParamName):\n",
    "        ConsentDF[SSIdParamName] = ConsentDF[ColumnToUseForSourceSubjectId]\n",
    "    \n",
    "    ConsentDF = ConsentDF.rename(columns = ColumnsToRename)    \n",
    "    ConsentDF = ConsentDF[[DbGapParamName, SubjectIdParamName, ConsentParamName, SSIdParamName]]\n",
    "    ConsentDF = ConsentDF.drop_duplicates()\n",
    "    print str(len(pd.unique(ConsentDF[SubjectIdParamName]))) + ' ' + StudyName + ' patients in consent file'\n",
    "    \n",
    "    return ConsentDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "##HELPER METHOD\n",
    "def GenerateHarmonizedDataFrameWithConsentInfo(HarmonizedDF, ConsentDF, InputDF):\n",
    "    ##INNER JOIN WITH SUBJECT TABLE(S) FROM dbGaP TO ADD CONSENT INFORMATION\n",
    "    HarmonizedConsentDF = pd.merge(HarmonizedDF, ConsentDF, on='SUBJECT_ID', how='inner')\n",
    "\n",
    "    print 'Found consent information for ' + str(len(pd.unique(HarmonizedConsentDF[SubjectIdParamName]))) + ' patients' \n",
    "    print 'Found Consent Levels: ' + str(pd.unique(HarmonizedConsentDF[ConsentParamName]))\n",
    "\n",
    "    HarmonizedConsentDF = HarmonizedConsentDF[['unique_subject_key', 'SUBJECT_ID', 'dbGaP_Subject_ID', 'topmed_study', 'CONSENT', \"SOURCE_SUBJECT_ID\"]]\n",
    "    print \n",
    "    \n",
    "    OutputDF = pd.concat([InputDF, HarmonizedConsentDF])\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## COPDGene Study\n",
    "def GetPatientsWithConsentInfoFromCopdGeneStudy(InputDF, HarmonizedDF):\n",
    "    print 'COPDGene..'\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR COPD GENE STUDY\n",
    "    CopdHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'COPDGene')\n",
    "\n",
    "    CopdConsentDF1 = ProcessConsentFile(CopdConsentFileName1, 'COPDGene')\n",
    "    CopdConsentDF2 = ProcessConsentFile(CopdConsentFileName2, 'COPDGene', SubjectIdParamName)\n",
    "\n",
    "    CopdConsentDF = pd.concat([CopdConsentDF1, CopdConsentDF2])\n",
    "\n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(CopdHarmonizedDF, CopdConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Framigham Heart Study\n",
    "def GetPatientsWithConsentInfoFromFraminghamHeartStudy(InputDF, HarmonizedDF):\n",
    "    print 'Framingham Heart Study..'\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR FRAMINGHAM STUDY\n",
    "    FhsHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'FHS', True)\n",
    "    \n",
    "    FhsConsentFile1 = baseDir + \"data_latest_version/\" + FhsConsentFileName1\n",
    "    FhsConsentDF1 = pd.read_csv(FhsConsentFile1, comment=\"#\", sep=\"\\t\")\n",
    "    FhsConsentDF1 = FhsConsentDF1[[DbGapParamName, 'consent_1217', 'source_subjid']]\n",
    "    FhsConsentDF1 = FhsConsentDF1.rename(columns={DbGapParamName:SubjectIdParamName, \"consent_1217\":ConsentParamName, \"source_subjid\":SSIdParamName})\n",
    "    FhsConsentDF1[DbGapParamName] = FhsConsentDF1[SubjectIdParamName]\n",
    "    FhsConsentDF1 = FhsConsentDF1.drop_duplicates()\n",
    "    print str(len(pd.unique(FhsConsentDF1[SubjectIdParamName]))) + ' FHS patients in first file'\n",
    "\n",
    "    FhsConsentFile2 = baseDir + 'data_latest_version/' + FhsConsentFileName2\n",
    "    FhsConsentDF2 = pd.read_csv(FhsConsentFile2, comment=\"#\", sep=\"\\t\")\n",
    "    FhsConsentDF2 = FhsConsentDF2[[DbGapParamName, ConsentParamName]]\n",
    "    FhsConsentDF2 = FhsConsentDF2.rename(columns={DbGapParamName:SubjectIdParamName})\n",
    "    FhsConsentDF2[DbGapParamName] = FhsConsentDF2[SubjectIdParamName]\n",
    "    FhsConsentDF2[SSIdParamName] = FhsConsentDF2[SubjectIdParamName]\n",
    "    FhsConsentDF2 = FhsConsentDF2.drop_duplicates()\n",
    "\n",
    "    print str(len(pd.unique(FhsConsentDF2[SubjectIdParamName]))) + ' FHS patients in Second TopMED file'\n",
    "\n",
    "    FhsConsentDF = pd.concat([FhsConsentDF1, FhsConsentDF2])\n",
    "    FhsConsentDF = FhsConsentDF.drop_duplicates()\n",
    "    print str(len(pd.unique(FhsConsentDF[SubjectIdParamName]))) + ' FHS patients in Merged Consent file'\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(FhsHarmonizedDF, FhsConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "##MESA Study\n",
    "def GetPatientsWithConsentInfoFromMesaStudy(InputDF, HarmonizedDF):\n",
    "    print 'MESA...'\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR MESA STUDY\n",
    "    MesaHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'MESA')\n",
    "    \n",
    "    ColumnsToRename={\"SUBJID\":SubjectIdParamName}\n",
    "    MesaConsentDF = ProcessConsentFileWithColumnRenaming(MesaConsentFileName, 'MESA', ColumnsToRename, 'SUBJID')\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(MesaHarmonizedDF, MesaConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Jackson Heart Study\n",
    "def GetPatientsWithConsentInfoFromJacksonHeartStudy(InputDF, HarmonizedDF):\n",
    "    print \"Jackson Heart Study..\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR JACKSON HEART STUDY\n",
    "    JhsHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'JHS')\n",
    "\n",
    "    JhsConsentFile1 = baseDir + 'data_latest_version/' + JhsConsentFileName1\n",
    "    JhsConsentDF1 = pd.read_csv(JhsConsentFile1, comment=\"#\", sep=\"\\t\")\n",
    "    JhsConsentDF1 = JhsConsentDF1[[DbGapParamName, 'SOURCE_SUBJECT_ID2', ConsentParamName]]\n",
    "    JhsConsentDF1 = JhsConsentDF1.rename(columns={\"SOURCE_SUBJECT_ID2\":SubjectIdParamName})\n",
    "    JhsConsentDF1[SSIdParamName] = JhsConsentDF1[SubjectIdParamName]\n",
    "    JhsConsentDF1 = JhsConsentDF1.drop_duplicates()\n",
    "    print str(len(pd.unique(JhsConsentDF1[SubjectIdParamName]))) + ' JHS patients in first file'\n",
    "\n",
    "    JhsConsentFile2 = baseDir + 'data_latest_version/' + JhsConsentFileName2\n",
    "    JhsConsentDF2 = pd.read_csv(JhsConsentFile2, comment=\"#\", sep=\"\\t\")\n",
    "    JhsConsentDF2 = JhsConsentDF2[[DbGapParamName, SSIdParamName, ConsentParamName]]\n",
    "    JhsConsentDF2[SubjectIdParamName] = JhsConsentDF2[SSIdParamName]\n",
    "    JhsConsentDF2 = JhsConsentDF2.drop_duplicates()\n",
    "    print str(len(pd.unique(JhsConsentDF2[SubjectIdParamName]))) + ' JHS patients in Second TopMED file'\n",
    "\n",
    "    JhsConsentDF = pd.concat([JhsConsentDF1, JhsConsentDF2])\n",
    "    JhsConsentDF.drop_duplicates()\n",
    "    print str(len(pd.unique(JhsConsentDF[SubjectIdParamName]))) + ' JHS patients in Merged Consent file'\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(JhsHarmonizedDF, JhsConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Amish Study\n",
    "def GetPatientsWithConsentInfoFromAmishStudy(InputDF, HarmonizedDF):\n",
    "    print \"Amish...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR AMISH STUDY\n",
    "    AmishHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'Amish')\n",
    "\n",
    "    AmishConsentDF = ProcessConsentFile(AmishConsentFileName, 'Amish')\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(AmishHarmonizedDF, AmishConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "##CHS Study\n",
    "def GetPatientsWithConsentInfoFromChsStudy(InputDF, HarmonizedDF):\n",
    "    print \"Cardiovascular Health Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR CHS STUDY\n",
    "    ChsHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'CHS')\n",
    "\n",
    "    ColumnsToRename={\"Individual_ID\":SubjectIdParamName, \"gencons\":ConsentParamName}\n",
    "    ChsConsentDF = ProcessConsentFileWithColumnRenaming(ChsConsentFileName, 'CHS', ColumnsToRename, 'Individual_ID')\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(ChsHarmonizedDF, ChsConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Women's Health Initiative Study\n",
    "def GetPatientsWithConsentInfoFromWhiStudy(InputDF, HarmonizedDF):\n",
    "    print \"Women\\'s Health Initiative Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR WHI STUDY\n",
    "    WhiHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'WHI')\n",
    "    \n",
    "    ColumnsToRename={\"SUBJID\":SubjectIdParamName}\n",
    "    WhiConsentDF = ProcessConsentFileWithColumnRenaming(WhiConsentFileName, 'WHI', ColumnsToRename, 'SUBJID')\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(WhiHarmonizedDF, WhiConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Heart And Vascular Study\n",
    "def GetPatientsWithConsentInfoFromHvhStudy(InputDF, HarmonizedDF):\n",
    "    print \"Heart and Vascular Health Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR HVH STUDY\n",
    "    HvhHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'HVH', True)\n",
    "    \n",
    "    HvhConsentFile1 = baseDir + \"data_latest_version/\" + HvhConsentFileName1 \n",
    "    HvhConsentDF1 = pd.read_csv(HvhConsentFile1, comment=\"#\", sep=\"\\t\")\n",
    "    ##The next two lines are needed because PANDAS makes row nums out of dbGaP_Subject_ID\n",
    "    HvhConsentDF1 = HvhConsentDF1.rename(columns={DbGapParamName:SubjectIdParamName, SubjectIdParamName:ConsentParamName, ConsentParamName:\"AFFECTION_STATUS\"})\n",
    "    HvhConsentDF1[DbGapParamName] = list(HvhConsentDF1.index)\n",
    "    HvhConsentDF1 = HvhConsentDF1[[DbGapParamName, SubjectIdParamName, ConsentParamName]]\n",
    "    HvhConsentDF1[SSIdParamName] = HvhConsentDF1[SubjectIdParamName]\n",
    "    HvhConsentDF1 = HvhConsentDF1.drop_duplicates()\n",
    "    print str(len(pd.unique(HvhConsentDF1[SubjectIdParamName]))) + ' HVH patients with consent in TopMED file'\n",
    "\n",
    "    HvhConsentFile2 = baseDir + \"data_latest_version/\" + HvhConsentFileName2\n",
    "    HvhConsentDF2 = pd.read_csv(HvhConsentFile2, comment=\"#\", sep=\"\\t\")\n",
    "    ##The next two lines are needed because PANDAS makes row nums out of dbGaP_Subject_ID\n",
    "    HvhConsentDF2 = HvhConsentDF2.rename(columns={DbGapParamName:SubjectIdParamName, SubjectIdParamName:ConsentParamName, ConsentParamName:\"AFFECTION_STATUS\"})\n",
    "    HvhConsentDF2[DbGapParamName] = list(HvhConsentDF2.index)\n",
    "    HvhConsentDF2 = HvhConsentDF2[[DbGapParamName, SubjectIdParamName, ConsentParamName]]\n",
    "    HvhConsentDF2[SSIdParamName] = HvhConsentDF2[SubjectIdParamName]\n",
    "    HvhConsentDF2 = HvhConsentDF2.drop_duplicates()\n",
    "    print str(len(pd.unique(HvhConsentDF2[SubjectIdParamName]))) + ' HVH patients with consent in second file'    \n",
    "      \n",
    "    HvhConsentDF = pd.concat([HvhConsentDF1, HvhConsentDF2])\n",
    "    HvhConsentDF = HvhConsentDF.drop_duplicates()\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(HvhHarmonizedDF, HvhConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Atherosclerosis Risk in Communities Study\n",
    "def GetPatientsWithConsentInfoFromAricStudy(InputDF, HarmonizedDF):\n",
    "    print \"Atherosclerosis Risk in Communities Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR ARIC STUDY\n",
    "    AricHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'ARIC')\n",
    "    \n",
    "    AricConsentDF1 = ProcessConsentFile(AricConsentFileName1, 'ARIC', SubjectIdParamName)\n",
    "    AricConsentDF2 = ProcessConsentFile(AricConsentFileName2, 'ARIC', SubjectIdParamName)\n",
    "    \n",
    "    AricConsentDF = pd.concat([AricConsentDF1, AricConsentDF2])\n",
    "    AricConsentDF = AricConsentDF.drop_duplicates()\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(AricHarmonizedDF, AricConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "##CARDIA Study\n",
    "def GetPatientsWithConsentInfoFromCardiaStudy(InputDF, HarmonizedDF):\n",
    "    print \"Coronary Artery Risk Development in Young Adults (CARDIA) Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR CARDIA STUDY\n",
    "    CardiaHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'CARDIA')\n",
    "    \n",
    "    ColumnsToRename = {\"SUBJID\":\"SUBJECT_ID\", \"Consent\":\"CONSENT\"}\n",
    "    CardiaConsentDF = ProcessConsentFileWithColumnRenaming(CardiaConsentFileName, 'CARDIA', ColumnsToRename, 'SOURCE_SUBJID')\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(CardiaHarmonizedDF, CardiaConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Cleveland Family Study\n",
    "def GetPatientsWithConsentInfoFromCfsStudy(InputDF, HarmonizedDF):\n",
    "    print \"Cleveland Family Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR CLEVELAND STUDY\n",
    "    CfsHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'CFS')\n",
    "\n",
    "    \n",
    "    ColumnsToRename = {\"SUBJID\":\"SUBJECT_ID\", \"Consent\":\"CONSENT\"}\n",
    "    CfsConsentDF = ProcessConsentFileWithColumnRenaming(CfsConsentFileName, 'CFS', ColumnsToRename)\n",
    "\n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(CfsHarmonizedDF, CfsConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "##GENEVA Genome-Wide Association Study of Venous Thrombosis (GWAS of VTE) Study\n",
    "def GetPatientsWithConsentInfoFromVteStudy(InputDF, HarmonizedDF):\n",
    "    print \"Venous Thromboembolism Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR VTE STUDY\n",
    "    VteHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'Mayo_VTE', True)\n",
    "\n",
    "    VteConsentFile = baseDir + \"data_latest_version/\" + VteConsentFileName\n",
    "    VteConsentDF = pd.read_csv(VteConsentFile, comment=\"#\", sep=\"\\t\")\n",
    "    VteConsentDF = VteConsentDF[['dbGaP SubjID', 'SUBJID', 'CONSENT', 'SOURCE_SUBJID']]\n",
    "    \n",
    "    VteConsentDF = VteConsentDF.rename(columns={\"SUBJID\":SubjectIdParamName, \"dbGaP SubjID\":DbGapParamName, \"SOURCE_SUBJID\":SSIdParamName}) \n",
    "    print str(len(pd.unique(VteConsentDF[SubjectIdParamName]))) + ' Mayo_VTE patients in file'\n",
    "        \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(VteHarmonizedDF, VteConsentDF, InputDF)\n",
    " \n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Epidemiology of Asthma in Costa Rica Study\n",
    "def GetPatientsWithConsentInfoFromCraStudy(InputDF, HarmonizedDF):\n",
    "    print \"Epidemiology of Asthma in Costa Rica  Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR CRA STUDY\n",
    "    CraHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'CRA')\n",
    "    \n",
    "    CraConsentDF = ProcessConsentFile(CraConsentFileName, 'CRA', SubjectIdParamName)\n",
    "       \n",
    "    CraConsentDF = CraConsentDF.drop_duplicates()\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(CraHarmonizedDF, CraConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Genetics of Lipid Lowering Drugs and Diet Network (GOLDN) Lipidomics\n",
    "def GetPatientsWithConsentInfoFromGoldnStudy(InputDF, HarmonizedDF):\n",
    "    print \"Genetics of Lipid Lowering Drugs and Diet Network (GOLDN) Lipidomics Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR GOLDN STUDY\n",
    "    GoldnHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'GOLDN', True)\n",
    "\n",
    "    GoldnConsentFile = baseDir + \"data_latest_version/\" + GoldnConsentFileName\n",
    "    GoldnConsentDF = pd.read_csv(GoldnConsentFile, comment=\"#\", sep=\"\\t\")\n",
    "    GoldnConsentDF = GoldnConsentDF[['dbGaP_Subject_ID', 'id', 'consent', 'subject_source_id']]\n",
    "    #because This consent file has id field instead of SUBJECT_ID\n",
    "    GoldnConsentDF = GoldnConsentDF.rename(columns={\"id\":SubjectIdParamName, \"consent\":ConsentParamName, \"subject_source_id\":SSIdParamName}) \n",
    "\n",
    "    print str(len(pd.unique(GoldnConsentDF[SubjectIdParamName]))) + ' GOLDN patients in file'\n",
    "        \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(GoldnHarmonizedDF, GoldnConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Hispanic Community Health Study Study of Latinos \n",
    "def GetPatientsWithConsentInfoFromHchsStudy(InputDF, HarmonizedDF):\n",
    "    print \"Hispanic Community Health Study - Study of Latinos...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR HCHS STUDY\n",
    "    HchsHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'HCHS_SOL')\n",
    "    \n",
    "    HchsConsentDF = ProcessConsentFile(HchsConsentFileName, 'HCHS_SOL')\n",
    "       \n",
    "    HchsConsentDF = HchsConsentDF.drop_duplicates()\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(HchsHarmonizedDF, HchsConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Study of Adiposity in Samoans Study\n",
    "def GetPatientsWithConsentInfoFromSasStudy(InputDF, HarmonizedDF):\n",
    "    print \"Study of Adiposity in Samoans Study...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR SAS STUDY\n",
    "    SasHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'SAS')\n",
    "    \n",
    "    SasConsentDF1 = ProcessConsentFile(SasConsentFileName1, 'SAS', SubjectIdParamName)\n",
    "    SasConsentDF2 = ProcessConsentFile(SasConsentFileName2, 'SAS', SubjectIdParamName)\n",
    "    \n",
    "    SasConsentDF = pd.concat([SasConsentDF1, SasConsentDF2])\n",
    "    SasConsentDF = SasConsentDF.drop_duplicates()\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(SasHarmonizedDF, SasConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Genetic Epidemiology Network of Arteriopathy (GENOA)\n",
    "def GetPatientsWithConsentInfoFromGenoaStudy(InputDF, HarmonizedDF):\n",
    "    print \"Genetic Epidemiology Network of Arteriopathy (GENOA)...\"\n",
    "    ##SELECT ROWS FROM HARMONIZED DF FOR Genetic Epidemiology Network of Arteriopathy (GENOA) STUDY\n",
    "    GenoaHarmonizedDF = GetStudySubsetFromHarmonizedData(HarmonizedDF, 'GENOA')\n",
    "    \n",
    "    GenoaConsentDF = ProcessConsentFile(GenoaConsentFileName, 'GENOA', SubjectIdParamName)\n",
    "    \n",
    "    OutputDF = GenerateHarmonizedDataFrameWithConsentInfo(GenoaHarmonizedDF, GenoaConsentDF, InputDF)\n",
    "    return OutputDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "##ASSEMBLES A DATA FRAME CONTAINING PATIENT NUMBERS AND CONSENT INFO FROM ALL PROJECTS\n",
    "def GetPatientsWithConsentInfoFromAllStudies(HarmonizedDF):\n",
    "    DF_I = GetPatientsWithConsentInfoFromCopdGeneStudy(StartDF, HarmonizedDF) ##StartDF is a Global Var\n",
    "    DF_II = GetPatientsWithConsentInfoFromFraminghamHeartStudy(DF_I, HarmonizedDF)\n",
    "    DF_III = GetPatientsWithConsentInfoFromMesaStudy(DF_II, HarmonizedDF)\n",
    "    DF_IV = GetPatientsWithConsentInfoFromJacksonHeartStudy(DF_III, HarmonizedDF)\n",
    "    DF_V = GetPatientsWithConsentInfoFromAmishStudy(DF_IV, HarmonizedDF)\n",
    "    DF_VI = GetPatientsWithConsentInfoFromChsStudy(DF_V, HarmonizedDF)\n",
    "    DF_VII = GetPatientsWithConsentInfoFromWhiStudy(DF_VI, HarmonizedDF)\n",
    "    DF_VIII = GetPatientsWithConsentInfoFromHvhStudy(DF_VII, HarmonizedDF)\n",
    "    DF_IX = GetPatientsWithConsentInfoFromAricStudy(DF_VIII, HarmonizedDF)\n",
    "    DF_X = GetPatientsWithConsentInfoFromCardiaStudy(DF_IX, HarmonizedDF)\n",
    "    DF_XI = GetPatientsWithConsentInfoFromCfsStudy(DF_X, HarmonizedDF)\n",
    "    DF_XII = GetPatientsWithConsentInfoFromVteStudy(DF_XI, HarmonizedDF)\n",
    "    DF_XIII = GetPatientsWithConsentInfoFromCraStudy(DF_XII, HarmonizedDF)\n",
    "    DF_XIV = GetPatientsWithConsentInfoFromGoldnStudy(DF_XIII, HarmonizedDF)\n",
    "    DF_XV = GetPatientsWithConsentInfoFromHchsStudy(DF_XIV, HarmonizedDF)\n",
    "    DF_XVI = GetPatientsWithConsentInfoFromSasStudy(DF_XV, HarmonizedDF)\n",
    "    DF_XVII = GetPatientsWithConsentInfoFromGenoaStudy(DF_XVI, HarmonizedDF)\n",
    "    \n",
    "    print 'DONE!!!'\n",
    "    \n",
    "    return DF_XVII"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "## THE MAIN METHOD\n",
    "def main():\n",
    "    print 'START..\\n'\n",
    "    HarmonizedDF = GenerateHarmonizedDF()\n",
    "    print str(len(pd.unique(HarmonizedDF['SUBJECT_ID']))) + ' patients found in harmonized dataset\\n'\n",
    "    \n",
    "    FinalDF = GetPatientsWithConsentInfoFromAllStudies(HarmonizedDF)\n",
    "\n",
    "    FinalDF.to_csv(r'./HarmonizedPatientsWithConsentInfo.csv', header=True, index=False)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "START..\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2878: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "229947 patients found in harmonized dataset\n",
      "\n",
      "COPDGene..\n",
      "10371 patients' data has been harmonized from COPDGene study\n",
      "10383 COPDGene patients in consent file\n",
      "10618 COPDGene patients in consent file\n",
      "Found consent information for 10371 patients\n",
      "Found Consent Levels: [1 2]\n",
      "\n",
      "Framingham Heart Study..\n",
      "15172 patients' data has been harmonized from FHS study\n",
      "18267 FHS patients in first file\n",
      "4154 FHS patients in Second TopMED file\n",
      "18267 FHS patients in Merged Consent file\n",
      "Found consent information for 6162 patients\n",
      "Found Consent Levels: [2 1 0]\n",
      "\n",
      "MESA...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python2.7/site-packages/ipykernel_launcher.py:25: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8296 patients' data has been harmonized from MESA study\n",
      "9780 MESA patients in consent file\n",
      "Found consent information for 8296 patients\n",
      "Found Consent Levels: [1 2]\n",
      "\n",
      "Jackson Heart Study..\n",
      "3602 patients' data has been harmonized from JHS study\n",
      "3597 JHS patients in first file\n",
      "5588 JHS patients in Second TopMED file\n",
      "5589 JHS patients in Merged Consent file\n",
      "Found consent information for 3596 patients\n",
      "Found Consent Levels: [3 4 2 1]\n",
      "\n",
      "Amish...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python2.7/site-packages/ipykernel_launcher.py:22: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1123 patients' data has been harmonized from Amish study\n",
      "2192 Amish patients in consent file\n",
      "Found consent information for 1123 patients\n",
      "Found Consent Levels: [2]\n",
      "\n",
      "Cardiovascular Health Study...\n",
      "5531 patients' data has been harmonized from CHS study\n",
      "5612 CHS patients in consent file\n",
      "Found consent information for 5531 patients\n",
      "Found Consent Levels: [1 2 3 4]\n",
      "\n",
      "Women's Health Initiative Study...\n",
      "143213 patients' data has been harmonized from WHI study\n",
      "143455 WHI patients in consent file\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python2.7/site-packages/ipykernel_launcher.py:7: DtypeWarning: Columns (1,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  import sys\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found consent information for 12141 patients\n",
      "Found Consent Levels: [1 2]\n",
      "\n",
      "Heart and Vascular Health Study...\n",
      "1204 patients' data has been harmonized from HVH study\n",
      "709 HVH patients with consent in TopMED file\n",
      "1204 HVH patients with consent in second file\n",
      "Found consent information for 1204 patients\n",
      "Found Consent Levels: [1 2]\n",
      "\n",
      "Atherosclerosis Risk in Communities Study...\n",
      "15676 patients' data has been harmonized from ARIC study\n",
      "15682 ARIC patients in consent file\n",
      "13546 ARIC patients in consent file\n",
      "Found consent information for 15676 patients\n",
      "Found Consent Levels: [1 2]\n",
      "\n",
      "Coronary Artery Risk Development in Young Adults (CARDIA) Study...\n",
      "3622 patients' data has been harmonized from CARDIA study\n",
      "3626 CARDIA patients in consent file\n",
      "Found consent information for 3622 patients\n",
      "Found Consent Levels: [1 2]\n",
      "\n",
      "Cleveland Family Study...\n",
      "1473 patients' data has been harmonized from CFS study\n",
      "994 CFS patients in consent file\n",
      "Found consent information for 971 patients\n",
      "Found Consent Levels: [1]\n",
      "\n",
      "Venous Thromboembolism Study...\n",
      "2935 patients' data has been harmonized from Mayo_VTE study\n",
      "2652 Mayo_VTE patients in file\n",
      "Found consent information for 2597 patients\n",
      "Found Consent Levels: [1]\n",
      "\n",
      "Epidemiology of Asthma in Costa Rica  Study...\n",
      "1533 patients' data has been harmonized from CRA study\n",
      "3789 CRA patients in consent file\n",
      "Found consent information for 1533 patients\n",
      "Found Consent Levels: [1]\n",
      "\n",
      "Genetics of Lipid Lowering Drugs and Diet Network (GOLDN) Lipidomics Study...\n",
      "968 patients' data has been harmonized from GOLDN study\n",
      "8544 GOLDN patients in file\n",
      "Found consent information for 968 patients\n",
      "Found Consent Levels: [1]\n",
      "\n",
      "Hispanic Community Health Study - Study of Latinos...\n",
      "12895 patients' data has been harmonized from HCHS_SOL study\n",
      "13296 HCHS_SOL patients in consent file\n",
      "Found consent information for 12895 patients\n",
      "Found Consent Levels: [2 1]\n",
      "\n",
      "Study of Adiposity in Samoans Study...\n",
      "3501 patients' data has been harmonized from SAS study\n",
      "3503 SAS patients in consent file\n",
      "1332 SAS patients in consent file\n",
      "Found consent information for 3501 patients\n",
      "Found Consent Levels: [1]\n",
      "\n",
      "Genetic Epidemiology Network of Arteriopathy (GENOA)...\n",
      "3462 patients' data has been harmonized from GENOA study\n",
      "6177 GENOA patients in consent file\n",
      "Found consent information for 3462 patients\n",
      "Found Consent Levels: [1]\n",
      "\n",
      "DONE!!!\n"
     ]
    }
   ],
   "source": [
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
